ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,309 @@
+  args.title = base::SysNSStringToUTF16(history.title);
+  args.floc_allowed = false;  // Disable Floc - Not allow tracking
+
+  history_service_->AddPage(args);
+}
+
+- (void)removeHistory:(IOSHistoryNode*)history {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes a specific URL using history service and web history service
+  history_service_->DeleteLocalAndRemoteUrl(web_history_service_,
+                                            net::GURLWithNSURL(history.url));
+}
+
+- (void)removeAllWithCompletion:(void (^)())completion {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes entire History and from all synced devices
+  history_service_->DeleteLocalAndRemoteHistoryBetween(
+      web_history_service_, base::Time::Min(), base::Time::Max(),
+      base::BindOnce(completion), &tracker_);
+}
+
+- (void)searchWithQuery:(NSString*)query
+               maxCount:(NSUInteger)maxCount
+             completion:(void (^)(NSArray<IOSHistoryNode*>* historyResults))
+                            completion {
+  __weak BraveHistoryAPI* weak_history_api = self;
+
+  auto search_with_query = ^(
+      NSString* query, NSUInteger maxCount,
+      void (^completion)(NSArray<IOSHistoryNode*>* historyResults)) {
+    BraveHistoryAPI* history_api = weak_history_api;
+    if (!history_api) {
+      completion(@[]);
+      return;
+    }
+
+    DCHECK(history_api->history_service_->backend_loaded());
+
+    // Check Query is empty for Fetching all history
+    // The entered query can be nil or empty String
+    BOOL fetchAllHistory = !query || [query length] == 0;
+    std::u16string queryString =
+        fetchAllHistory ? std::u16string() : base::SysNSStringToUTF16(query);
+
+    // Creating fetch options for querying history
+    history::QueryOptions options;
+    options.duplicate_policy =
+        fetchAllHistory ? history::QueryOptions::REMOVE_DUPLICATES_PER_DAY
+                        : history::QueryOptions::REMOVE_ALL_DUPLICATES;
+    options.max_count = fetchAllHistory ? 0 : static_cast<int>(maxCount);
+    options.matching_algorithm =
+        query_parser::MatchingAlgorithm::ALWAYS_PREFIX_SEARCH;
+
+    history_api->_query_completion = completion;
+
+    history_api->browsing_history_service_->QueryHistory(queryString, options);
+  };
+
+  base::PostTask(
+      FROM_HERE, {web::WebThread::UI},
+      base::BindOnce(search_with_query, query, maxCount, completion));
+}
+
+#pragma mark - BraveHistoryDriverDelegate
+
+- (void)
+    historyQueryWasCompletedWithResults:
+        (const std::vector<history::BrowsingHistoryService::HistoryEntry>&)
+            results
+                       queryResultsInfo:(const history::BrowsingHistoryService::
+                                             QueryResultsInfo&)queryResultsInfo
+                    continuationClosure:(base::OnceClosure)continuationClosure {
+  if (_query_completion) {
```

Comment:
seems like we also need to handle this? https://source.chromium.org/chromium/chromium/src/+/main:ios/chrome/browser/ui/history/history_table_view_controller.mm;l=288?q=historyQueryWasCompletedWithResults&ss=chromium

Reply:
Pagination is removed.
Same with https://github.com/brave/brave-core/pull/8869#discussion_r646984276